File: QuickSort/quick_sort.cpp
#include <iostream>
#include <random>
#include <chrono>

void swap(int* a, int* b)
{
    int t = *a;
    *a = *b;
    *b = t;
}

int partition (int arr[], int low, int high)
{
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j <= high- 1; j++)
    {
        if (arr[j] <= pivot)
        {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

void quickSort(int arr[], int low, int high)
{
    if (low < high)
    {
        int pivot = partition(arr, low, high);
        quickSort(arr, low, pivot - 1);
        quickSort(arr, pivot + 1, high);
    }
}

int main()
{
    std::cout << "Sorting arrays of size 10,000,000\n";
    int total_time = 0;
    for(int i = 0; i < 25; i++)
    {
        const int size = 10000000;
        int *array = new int[size];
        std::random_device rd;
        std::mt19937 mt(rd());
        std::uniform_real_distribution<double> dist(0, size);

        for(int j = 0; j < size; j++){
            array[j] = dist(mt);
        }
        auto start = std::chrono::high_resolution_clock::now();
        quickSort(array, 0, size - 1);
        auto finish = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(finish - start);
        total_time += duration.count();
        std::cout << "\tSorted array " << i << ", took "<< duration.count() << "ms" << std::endl;
        delete [] array;
    }
    std::cout << "\t------------------------------\n" << "\tAverage: " << total_time / 25 << "ms\n\n";
}

File: QuickSort/quick_sort.rs
use std::time::{Duration, Instant};

fn run_benchmark(size: u32) -> Duration{
    let mut vec = make_rand_vec(size);
    let now = Instant::now();
    quick_sort(&mut vec);
    return now.elapsed()
}

fn quick_sort(array: &mut [i32]) {
    if !array.is_empty() {
        let partition_index = partition(array);
        let len = array.len();
        quick_sort(&mut array[0..partition_index]);
        quick_sort(&mut array[partition_index + 1..len]);
    }
}

fn partition(array: &mut [i32]) -> usize {
    let len = array.len();
    let pivot = array[len - 1];
    let mut i = 0;
    let mut j = 0;
    while j < len - 1 {
        if array[j] <= pivot {
            array.swap(i, j);
            i += 1;
        }
        j += 1;
    }
    array.swap(i, len - 1);
    return i
}

fn make_rand_vec(size: u32) -> Vec<i32> {
    let mut array: Vec<i32> = Vec::new();
    for _ in 0..size {
        array.push(rand::random::<i32>());
    }
    return array
}

fn main() {
    let size = 10000000;
    println!("\nSorting arrays of size {}", size);
    let mut total = 0;
    for i in 0..25{
        let time = run_benchmark(size);
        total += time.as_millis();
        println!("\tSorted array {}, took {}ms", i, time.as_millis());
    }
    println!("\t-----------------------------");
    println!("\tAverage: {}ms", total/25);
}